core/src/main/java/com/github/mizool/core/concurrent/FutureStreamJoiner.java

```diff
@@ -0,0 +1,195 @@
+         * <br>
+         * See {@link FutureStreamJoiner} for details.
+         *
+         * @param listenableFutures the futures to join
+         * @param concurrencyLimit the maximum number of incomplete futures at any point in time
+         * @param executorService where to execute the stream consumer
+         */
+        public ListenableFuture<Void> join(
+            Stream<ListenableFuture<Void>> listenableFutures, int concurrencyLimit, ExecutorService executorService)
+        {
+            verifyConcurrencyLimit(concurrencyLimit);
+
+            Stream<Void> results = BufferedStreamAdapter.listenable()
+                .adapt(listenableFutures, concurrencyLimit, executorService);
+
+            return (ListenableFuture<Void>) MoreExecutors.listeningDecorator(executorService)
+                .submit(consumeStream(results));
+        }
+    }
+
+    /**
+     * Provides fluent syntax for joining completable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     */
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
+    public class Completable
+    {
+        /**
+         * Returns a future that will complete normally once all futures in the given stream have completed
+         * normally. <br>
+         * <br>
+         * See {@link FutureStreamJoiner} for details.
+         *
+         * @param completableFutures the futures to join
+         * @param concurrencyLimit the maximum number of incomplete futures at any point in time
+         * @param executorService where to execute the stream consumer
+         */
+        public CompletableFuture<Void> join(
+            Stream<CompletableFuture<Void>> completableFutures, int concurrencyLimit, ExecutorService executorService)
+        {
+            verifyConcurrencyLimit(concurrencyLimit);
+
+            Stream<Void> results = BufferedStreamAdapter.completable()
+                .adapt(completableFutures, concurrencyLimit, executorService);
+
+            return CompletableFuture.runAsync(consumeStream(results), executorService);
+        }
+    }
+
+    /**
+     * Entry point for joining listenable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     *
+     * @return fluent syntax
+     */
+    public Listenable listenable()
+    {
+        return new Listenable();
+    }
+
+    /**
+     * Entry point for joining completable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     *
+     * @return fluent syntax
+     */
+    public Completable completable()
+    {
+        return new Completable();
+    }
+
+    private void verifyConcurrencyLimit(int concurrencyLimit)
+    {
+        if (concurrencyLimit <= 0)
+        {
+            throw new IllegalArgumentException("Concurrency limit must be greater than 0");
+        }
+    }
+
+    private Runnable consumeStream(Stream<Void> stream)
+    {
+        return new StreamConsumingWorker(stream);
+    }
+
+    @Value
```

Comment:
See my comment on the `ListenableFutureCallback` in the `BufferedStreamAdapter` on why I think `@Value` should not be used here.

Replace both annotations with a `@RequiredArgsConstructor` and make the field `private final`.